src/main/java/org/veupathdb/service/eda/ds/plugin/CategoricalCountsPlugin.java

```diff
@@ -0,0 +1,140 @@
+      new ManagedMap<>(5000, 2000);
+
+  private String _cacheKey;
+  private CategoricalCountsPostResponse _cachedResponse;
+
+  @Override
+  protected Class<CategoricalCountsPostRequest> getVisualizationRequestClass() {
+    return CategoricalCountsPostRequest.class;
+  }
+
+  @Override
+  protected Class<CategoricalCountsSpec> getVisualizationSpecClass() {
+    return CategoricalCountsSpec.class;
+  }
+
+  @Override
+  public ConstraintSpec getConstraintSpec() {
+    return new ConstraintSpec()
+      .dependencyOrder(List.of("variable"))
+      .pattern()
+        .element("variable")
+          .shapes(APIVariableDataShape.CATEGORICAL, APIVariableDataShape.ORDINAL)
+          .description("Variable must be categorical.")
+      .done();
+  }
+  
+  @Override
+  protected void validateVisualizationSpec(CategoricalCountsSpec pluginSpec) throws ValidationException {
+    validateInputs(new DataElementSet()
+      .var("variable", pluginSpec.getVariable()));
+  }
+
+  @Override
+  protected List<StreamSpec> getRequestedStreams(CategoricalCountsSpec pluginSpec) {
+    String entityId = pluginSpec.getVariable().getEntityId();
+    _cacheKey = EncryptionUtil.md5(
+        JsonUtil.serializeObject(getSubsetFilters()) +
+        "|" + JsonUtil.serializeObject(pluginSpec)
+    );
+    _cachedResponse = RESULT_CACHE.get(_cacheKey);
+    if (_cachedResponse != null) LOG.info("Found cached result.");
+
+    return _cachedResponse != null ? Collections.emptyList() : ListBuilder.asList(
+      new StreamSpec(DEFAULT_SINGLE_STREAM_NAME, pluginSpec.getVariable().getEntityId())
+        .addVar(pluginSpec.getVariable()));
+  }
+
+  @Override
+  protected void writeResults(OutputStream out, Map<String, InputStream> dataStreams) throws IOException {
+    CategoricalCountsSpec spec = getPluginSpec();
+
+    // create scanner and line parser
+    InputStreamReader isReader = new InputStreamReader(new BufferedInputStream(dataStreams.get(DEFAULT_SINGLE_STREAM_NAME)));
+    BufferedReader reader = new BufferedReader(isReader);
+    DelimitedDataParser parser = new DelimitedDataParser(reader.readLine(), TAB, true);
+
+    // establish column header indexes
+    Function<VariableSpec,Integer> indexOf = var -> parser.indexOfColumn(getUtil().toColNameOrEmpty(var)).orElseThrow();
+    int varIndex = indexOf.apply(spec.getVariable());
+
+    // this just needs to make a hashmap of values and their counts
+    Map<String, MutableInt> categoryCounts = new HashMap<>();
+    String nextLine = reader.readLine();
+
+    while (nextLine != null) {
+      String[] row = parser.parseLineToArray(nextLine);
+
+      categoryCounts.putIfAbsent(row[varIndex], new MutableInt());
+      categoryCounts.get(row[varIndex]).increment();
+
+      nextLine = reader.readLine();
+    }
+
+    // begin output object and single property containing array of elements
+    out.write("{\"categories\":[".getBytes(StandardCharsets.UTF_8));
```

Comment:
Looks like this just creates a histogram of a categorical var for a given subset.  If that's all we need, I could probably just expose a call to the subsetting service's distribution endpoint (with subset of your choice).  You'd have to parse that endpoint's API but it should be a lot quicker than this.  Only downside I can think of is that you couldn't use a derived var (for now).  Is there a reason this wouldn't be better?

Reply:
p.s. Dan Galdi is beginning to think about changing over the count and distribution endpoints to use his much faster file-based subsetting system.  Probably still not as fast as a cached value, but probably pretty quick.

Reply:
i think i like this idea. curious if @bobular sees any reason to not do this?